src/main/scala/org/wikimedia/analytics/hdfstools/HdfsRsyncConfig.scala

```diff
@@ -0,0 +1,360 @@
+                (if (deleteExcluded && !deleteExtraneous) Seq("delete-excluded must be used in conjunction with delete") else Nil)
+        }
+
+        prepareErrorMessage(errorMessages, "Error validating flags:")
+    }
+
+    /**
+     * Validate the whole config at once.
+     * This allows for the scopt parser to be provided with all error messages at once
+     * instead of failing at the first one.
+     *
+     * The error messages concatenation is done by folding validation functions results,
+     * accumulating errors as Option[String]. The fold is initialized with None(success),
+     * and at each stage checks for the function result. It returns an updated (error)
+     * status if the function result is an error.
+     *
+     * @return None if validation succeeds, Some(error-message) otherwise.
+     */
+    def validate: Option[String] = {
+        Seq(
+            validateSrc,
+            validateDst,
+            validateChmods,
+            validateFilterRules,
+            validateFlags
+        ).fold(None)((accumulatedValidationResults, newValidationResult) => {
+            // newValidationResult is a success, return unchanged accumulatedValidationResults
+            if (newValidationResult.isEmpty) {
+                accumulatedValidationResults
+            } else {
+                if (accumulatedValidationResults.isEmpty) {
+                    // First error we encounter (accumulatedValidationResults would not be empty otherwise)
+                    // Add error message header to newValidationResult message
+                    Some(s"Argument parsing error:\n\t${newValidationResult.get}")
+                } else {
+                    // there is already existing error(s) and a new one - Concatenate them
+                    Some(s"${accumulatedValidationResults.get}\n\t${newValidationResult.get}")
+                }
+            }
+        })
+    }
+
+    /********************************************************************************
+     * Functions and classes needed for config initialization of internals
+     */
+
+    /**
+     * Create an optional HDFS [[ChmodParser]] from a rebuilt chmod string (comma-separated
+     * chmod commands) for either files or dirs. The rebuilt command string is extracted from
+     * the fully validated list of chmod commands as follow: the full command list is fold,
+     * concatenating in an option the files or directory commands with the 'F' or 'D' prefix
+     * removed if any.
+     * The ChmodParser is then created if the fold has generated some command.
+     *
+     * @param chmods the validated sequence of commands containing both files and dirs commands
+     * @param files whether the generated command string should be for files or for dirs
+     * @return the generated comma-separated chmod command for either files or dirs, if any
+     */
+    private def getChmodParser(chmods: Seq[String], files: Boolean): Option[ChmodParser] = {
```

Comment:
This function is pretty confusing.  Can you just pre-filter for the type of chmod commands you want and then parse them?
```
val directoryChmodCommands = getChmodParser(chmodCommands.filter(!_.startsWith('F')))
val fileChmodCommands = getChmodParser(chmodCommands.filter(!_.startsWith('D')))
```

Reply:
done! Looks better indeed :)